File: Data Structures & Algorithms/validate-parentheses/submission-1.ts
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s: string): boolean {
        const CLOSED_BRACKETS = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        const stack = []
        let i = 0
        let n = s.length


        while (i < n) {
            if (CLOSED_BRACKETS[s[i]]) {
                if (stack[stack.length - 1] === CLOSED_BRACKETS[s[i]]) {
                    stack.pop()
                    i++
                    continue
                } else {
                    return false
                }
            }

            stack.push(s[i])
            i++
        }


        return stack.length === 0

    }
}

File: Data Structures & Algorithms/next-greater-element-i/submission-1.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack: List[int] = []
        index_map = {}
        res = []
        
        for num in nums2:
            while len(stack) > 0 and num > stack[-1]:
                index_map[stack.pop()] = num

            stack.append(num)
        
        for num in nums1:
            if num in index_map:
                res.append(index_map[num])
            else:
                res.append(-1)

        return res